src/Microsoft.ServiceFabric.Actors.KVSToRCMigration/DataValidationWorker.cs

```diff
@@ -0,0 +1,201 @@
+
+    internal class DataValidationWorker : WorkerBase
+    {
+        private static readonly string TraceType = typeof(DataValidationWorker).Name;
+        private StatefulServiceInitializationParameters initParams;
+        private ServicePartitionClient<HttpCommunicationClient> servicePartitionClient;
+        private MigrationSettings migrationSettings;
+
+        public DataValidationWorker(
+            KVStoRCMigrationActorStateProvider stateProvider,
+            ActorTypeInformation actorTypeInfo,
+            ServicePartitionClient<HttpCommunicationClient> servicePartitionClient,
+            MigrationSettings migrationSettings,
+            WorkerInput workerInput,
+            string traceId)
+            : base(stateProvider, workerInput, traceId)
+        {
+            this.initParams = this.StateProvider.GetInitParams();
+            this.servicePartitionClient = servicePartitionClient;
+            this.migrationSettings = migrationSettings;
+        }
+
+        public override async Task<WorkerResult> StartWorkAsync(CancellationToken cancellationToken)
+        {
+            ActorTrace.Source.WriteInfoWithId(
+                        TraceType,
+                        this.TraceId,
+                        $"Starting or resuming data validation worker\n Input: {this.Input.ToString()}");
+
+            try
+            {
+                var startSN = this.Input.StartSeqNum;
+                var endSN = this.Input.EndSeqNum;
+
+                var migratedKeys = new List<string>();
+                using (var tx = this.StateProvider.GetStateManager().CreateTransaction())
+                {
+                    var savedMigratedKeys = await this.MetadataDict.TryGetValueAsync(tx, MigrationConstants.MigrationKeysMigrated);
+                    if (savedMigratedKeys.HasValue)
+                    {
+                        migratedKeys = savedMigratedKeys.Value.Split(DefaultDelimiter).ToList();
+                    }
+                }
+
+                var noOfKeysToValidate = Math.Round(migratedKeys.Count * (this.migrationSettings.PercentageOfMigratedDataToValidate / 100), 0);
+                var rand = new Random();
+                var validatedKeysIndices = new List<int>();
+
+                for (long index = startSN; index <= endSN; index++)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    if (this.migrationSettings.PercentageOfMigratedDataToValidate <= 0)
+                    {
+                        break;
+                    }
+
+                    var tIndex = rand.Next(((int)startSN), ((int)endSN));
+                    if (validatedKeysIndices.Contains(tIndex))
+                    {
+                        // skip index already validated
+                        continue;
+                    }
+
+                    if (await this.GetDataFromKvsAndValidateWithRCAsync(migratedKeys[(int)index], this.Input.WorkerId, cancellationToken))
```

Comment:
GetDataFromKvsAndValidateWithRCAsync(migratedKeys[(int)**_tIndex_**], this.Input.WorkerId, cancellationToken)